File: backend/app/api/v1/alerts.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload
from typing import List, Optional, Dict, Any
from datetime import datetime, timezone
import logging
from app.db.session import get_db
from app.models.alert import Alert
from app.models.alert_trigger import AlertTrigger
from app.models.symbol import Symbol
from app.schemas.alert import AlertCreate, AlertUpdate, AlertResponse, AlertTrigger as AlertTriggerSchema
from app.core.enums import AlertCondition
from app.services.indicator_registry.registry import get_registry
from app.api.decorators import public_endpoint

logger = logging.getLogger(__name__)
router = APIRouter()


async def _add_alert_label(alert: Alert, response: dict) -> dict:
    """Add computed alert_label to response dictionary."""
    response['alert_label'] = alert.alert_label
    return response


async def _add_trigger_alert_labels(triggers: List[AlertTrigger], db: AsyncSession) -> List[dict]:
    """Add alert_label to trigger events by joining with alerts."""
    result = []
    for trigger in triggers:
        trigger_dict = {
            'id': trigger.id,
            'alert_id': trigger.alert_id,
            'triggered_at': trigger.triggered_at,
            'observed_price': trigger.observed_price,
            'indicator_value': trigger.indicator_value,
            'trigger_type': trigger.trigger_type,
            'trigger_message': trigger.trigger_message,
            'delivery_status': trigger.delivery_status,
            'retry_count': trigger.retry_count,
            'last_retry_at': trigger.last_retry_at,
            'alert_label': None  # Will be populated below
        }
        # Fetch alert to compute label
        alert_result = await db.execute(select(Alert).where(Alert.id == trigger.alert_id))
        alert = alert_result.scalars().first()
        if alert:
            trigger_dict['alert_label'] = alert.alert_label
        result.append(trigger_dict)
    return result


@router.post("/", response_model=AlertResponse, status_code=201)
@public_endpoint
async def create_alert(
    alert_in: AlertCreate,
    db: AsyncSession = Depends(get_db)
):
    """
    Create a new price alert or indicator-based alert.

    Supports both symbol_id (int) and symbol (string) - if symbol is provided,
    it will be resolved to symbol_id automatically.

    Price Conditions:
    - above: Triggers when current > threshold AND previous <= threshold
    - below: Triggers when current < threshold AND previous >= threshold
    - crosses_up: Triggers when previous < threshold AND current >= threshold
    - crosses_down: Triggers when previous > threshold AND current <= threshold

    Indicator Conditions (cRSI only in initial implementation):
    - indicator_crosses_upper: Triggers when indicator crosses above threshold (upper band)
    - indicator_crosses_lower: Triggers when indicator crosses below threshold (lower band)

    For indicator-based alerts:
    - messages: Flexible map of condition_type to message (e.g., {"indicator_crosses_upper": "Sell!"})
    - message_upper/message_lower: Legacy cRSI format (for backward compatibility)
    - enabled_conditions: Which conditions are enabled (e.g., {"upper": true, "lower": true})
    - cooldown: Minimum 5 seconds, default 60 seconds
    """
    # Validate condition is a valid AlertCondition
    try:
        AlertCondition(alert_in.condition)
    except ValueError:
        logger.warning(f"Invalid condition: {alert_in.condition}")
        raise HTTPException(
            status_code=422,
            detail=f"Invalid condition '{alert_in.condition}'. Must be one of: {[c.value for c in AlertCondition]}"
        )

    # Log the incoming data for debugging
    logger.info(f"Creating alert: symbol={alert_in.symbol}, condition={alert_in.condition}, indicator_name={alert_in.indicator_name}")

    # Resolve symbol to symbol_id if symbol string is provided
    symbol_id = alert_in.symbol_id
    symbol_ticker = None
    if symbol_id is None:
        # Try to resolve symbol string to symbol_id
        if alert_in.symbol:
            symbol_result = await db.execute(select(Symbol).where(Symbol.ticker == alert_in.symbol.upper()))
            symbol_obj = symbol_result.scalars().first()
            if symbol_obj:
                symbol_id = symbol_obj.id
                symbol_ticker = symbol_obj.ticker
            else:
                raise HTTPException(
                    status_code=404,
                    detail=f"Symbol '{alert_in.symbol}' not found"
                )
        else:
            raise HTTPException(
                status_code=422,
                detail="Either symbol_id or symbol must be provided"
            )

    alert = Alert(
        symbol_id=symbol_id,
        condition=alert_in.condition.value if isinstance(alert_in.condition, AlertCondition) else alert_in.condition,
        threshold=alert_in.threshold,
        is_active=alert_in.is_active,
        interval=alert_in.interval,
        cooldown=alert_in.cooldown,
        # Indicator fields
        indicator_name=alert_in.indicator_name,
        indicator_field=alert_in.indicator_field,
        indicator_params=alert_in.indicator_params,
        # Message fields - use messages if provided, otherwise fall back to legacy format
        messages=alert_in.messages,
        message_upper=alert_in.message_upper,
        message_lower=alert_in.message_lower,
        enabled_conditions=alert_in.enabled_conditions
    )
    db.add(alert)
    await db.commit()
    await db.refresh(alert)

    # Return dict with symbol field populated (to avoid lazy loading issues)
    alert_dict = {**alert.__dict__, 'symbol': symbol_ticker}
    return alert_dict


@router.get("/", response_model=List[AlertResponse])
@public_endpoint
async def list_alerts(
    symbol_id: Optional[int] = Query(None, description="Filter by symbol ID"),
    symbol: Optional[str] = Query(None, description="Filter by symbol (preferred, simpler frontend)"),
    is_active: Optional[bool] = Query(None, description="Filter by active status"),
    db: AsyncSession = Depends(get_db)
):
    """
    List all alerts, optionally filtered by symbol and active status.

    Note: symbol parameter (string) is preferred over symbol_id for simpler frontend integration.
    """
    query = select(Alert).options(selectinload(Alert.symbol_obj))

    # Support both symbol_id (int) and symbol (string) filtering
    if symbol_id is not None:
        query = query.where(Alert.symbol_id == symbol_id)
    elif symbol is not None:
        # Resolve symbol string to symbol_id
        symbol_result = await db.execute(select(Symbol).where(Symbol.ticker == symbol.upper()))
        symbol_obj = symbol_result.scalars().first()
        if symbol_obj:
            query = query.where(Alert.symbol_id == symbol_obj.id)
        else:
            # Symbol not found, return empty list
            return []

    if is_active is not None:
        query = query.where(Alert.is_active == is_active)

    query = query.order_by(Alert.created_at.desc())
    result = await db.execute(query)
    alerts = result.scalars().all()

    # Add symbol to each alert response (symbol_obj is now eagerly loaded)
    response_alerts = []
    for alert in alerts:
        alert_dict = {
            **alert.__dict__,
            'symbol': alert.symbol_obj.ticker if hasattr(alert, 'symbol_obj') and alert.symbol_obj else None
        }
        response_alerts.append(alert_dict)

    return response_alerts


@router.get("/indicator-conditions")
@public_endpoint
async def get_indicator_conditions(
    indicator_name: str = Query(..., description="Name of the indicator")
) -> Dict[str, Any]:
    """
    Get available alert conditions for a specific indicator.

    Returns the alert templates that define which conditions can be used
    with this indicator (e.g., crosses_upper, turns_positive, slope_bullish).

    Each condition includes:
    - condition_type: The alert condition enum value
    - label: Human-readable name for the condition
    - description: What the condition does
    - applicable_fields: Which indicator fields this condition applies to
    - requires_threshold: Whether this condition needs a threshold value
    """
    registry = get_registry()
    indicator = registry.get_by_base_name(indicator_name)

    if indicator is None:
        raise HTTPException(
            status_code=404,
            detail=f"Indicator '{indicator_name}' not found. Available indicators: {[ind.base_name for ind in registry._indicators.values()]}"
        )

    # Convert alert templates to response format
    conditions = [
        {
            "condition_type": template.condition_type,
            "label": template.label,
            "description": template.description,
            "requires_threshold": template.requires_threshold,
            "applicable_fields": template.applicable_fields
        }
        for template in indicator.alert_templates
    ]

    return {
        "indicator_name": indicator_name,
        "conditions": conditions
    }


@router.get("/triggers/recent", response_model=List[AlertTriggerSchema])
@public_endpoint
async def get_recent_triggers(
    symbol_id: Optional[int] = Query(None, description="Filter by symbol ID"),
    symbol: Optional[str] = Query(None, description="Filter by symbol (preferred, simpler frontend)"),
    limit: int = Query(500, ge=1, le=1000, description="Maximum number of triggers to return"),
    offset: int = Query(0, ge=0, description="Number of triggers to skip (for pagination)"),
    db: AsyncSession = Depends(get_db)
):
    """
    Get recent alert triggers across all alerts for the global Log tab.

    Returns triggers in reverse chronological order (newest first).
    Optionally filter by symbol to see triggers for specific symbols.

    Response includes:
    - trigger_type: Which condition fired ("upper" or "lower")
    - trigger_message: The direction-specific message used
    - alert_label: Computed label for display
    """
    query = select(AlertTrigger)

    # Support both symbol_id (int) and symbol (string) filtering
    if symbol_id is not None:
        query = query.join(Alert, AlertTrigger.alert_id == Alert.id).where(Alert.symbol_id == symbol_id)
    elif symbol is not None:
        # Resolve symbol string to symbol_id
        symbol_result = await db.execute(select(Symbol).where(Symbol.ticker == symbol.upper()))
        symbol_obj = symbol_result.scalars().first()
        if symbol_obj:
            query = query.join(Alert, AlertTrigger.alert_id == Alert.id).where(Alert.symbol_id == symbol_obj.id)
        else:
            # Symbol not found, return empty list
            return []

    query = query.order_by(AlertTrigger.triggered_at.desc()).limit(limit).offset(offset)

    result = await db.execute(query)
    triggers = result.scalars().all()

    # trigger_type and trigger_message are now stored in the database
    # alert_label will be computed by Pydantic from the Alert model relationship
    return triggers


@router.get("/{alert_id}", response_model=AlertResponse)
@public_endpoint
async def get_alert(
    alert_id: int,
    db: AsyncSession = Depends(get_db)
):
    """Get a specific alert by ID."""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalars().first()

    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    return alert


@router.put("/{alert_id}", response_model=AlertResponse)
@public_endpoint
async def update_alert(
    alert_id: int,
    alert_in: AlertUpdate,
    db: AsyncSession = Depends(get_db)
):
    """Update an existing alert."""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalars().first()

    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    # Validate condition if provided
    if alert_in.condition is not None:
        try:
            AlertCondition(alert_in.condition)
            alert.condition = alert_in.condition.value if isinstance(alert_in.condition, AlertCondition) else alert_in.condition
        except ValueError:
            raise HTTPException(
                status_code=422,
                detail=f"Invalid condition '{alert_in.condition}'. Must be one of: {[c.value for c in AlertCondition]}"
            )

    # Update fields
    if alert_in.threshold is not None:
        alert.threshold = alert_in.threshold
    if alert_in.is_active is not None:
        alert.is_active = alert_in.is_active
    if alert_in.cooldown is not None:
        alert.cooldown = alert_in.cooldown
    if alert_in.message_upper is not None:
        alert.message_upper = alert_in.message_upper
    if alert_in.message_lower is not None:
        alert.message_lower = alert_in.message_lower
    if alert_in.enabled_conditions is not None:
        alert.enabled_conditions = alert_in.enabled_conditions
    if alert_in.indicator_params is not None:
        alert.indicator_params = alert_in.indicator_params

    await db.commit()
    await db.refresh(alert)
    return alert


@router.delete("/{alert_id}", status_code=204)
@public_endpoint
async def delete_alert(
    alert_id: int,
    db: AsyncSession = Depends(get_db)
):
    """Delete an alert and all its trigger history (cascade delete)."""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalars().first()

    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    await db.delete(alert)
    await db.commit()
    return None


@router.post("/{alert_id}/mute", response_model=AlertResponse)
@public_endpoint
async def mute_alert(
    alert_id: int,
    db: AsyncSession = Depends(get_db)
):
    """Mute an alert (prevent it from creating trigger events)."""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalars().first()

    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    alert.is_active = False
    await db.commit()
    await db.refresh(alert)
    return alert


@router.post("/{alert_id}/unmute", response_model=AlertResponse)
@public_endpoint
async def unmute_alert(
    alert_id: int,
    db: AsyncSession = Depends(get_db)
):
    """Unmute an alert (allow it to create trigger events again)."""
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalars().first()

    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    alert.is_active = True
    await db.commit()
    await db.refresh(alert)
    return alert


@router.get("/{alert_id}/triggers", response_model=List[AlertTriggerSchema])
@public_endpoint
async def get_alert_triggers(
    alert_id: int,
    limit: int = Query(100, ge=1, le=1000, description="Maximum number of triggers to return"),
    db: AsyncSession = Depends(get_db)
):
    """
    Get trigger history for a specific alert.

    Returns a list of when this alert has fired, including:
    - trigger_type: Which condition fired ("upper" or "lower")
    - trigger_message: The direction-specific message used
    - alert_label: Computed label for display
    """
    # Verify alert exists
    result = await db.execute(select(Alert).where(Alert.id == alert_id))
    alert = result.scalars().first()
    if not alert:
        raise HTTPException(status_code=404, detail="Alert not found")

    # Get triggers
    query = select(AlertTrigger).where(
        AlertTrigger.alert_id == alert_id
    ).order_by(
        AlertTrigger.triggered_at.desc()
    ).limit(limit)

    result = await db.execute(query)
    triggers = result.scalars().all()

    # trigger_type and trigger_message are now stored in the database
    # alert_label will be computed by Pydantic from the Alert model
    return triggers


@router.delete("/triggers/{trigger_id}", status_code=204)
@public_endpoint
async def delete_trigger(
    trigger_id: int,
    db: AsyncSession = Depends(get_db)
):
    """
    Delete a specific alert trigger (log entry).

    This allows users to remove individual trigger entries from the log,
    useful for cleaning up orphaned or unwanted trigger records.
    """
    result = await db.execute(select(AlertTrigger).where(AlertTrigger.id == trigger_id))
    trigger = result.scalars().first()

    if not trigger:
        raise HTTPException(status_code=404, detail="Trigger not found")

    await db.delete(trigger)
    await db.commit()
    return None

File: frontend/src/api/alerts.ts
import client from './client'
import type { GuestAlert, AlertCondition as GuestAlertCondition } from '../types/auth'
import type { AlertNotificationSettingsUpdate } from '../types/notification'
import { createAuthenticatedAxios } from '@/services/authService'

export type AlertCondition =
  | 'above'
  | 'below'
  | 'crosses_up'
  | 'crosses_down'
  | 'indicator_above_upper'
  | 'indicator_below_lower'
  | 'indicator_crosses_upper'
  | 'indicator_crosses_lower'
  | 'indicator_turns_positive'
  | 'indicator_turns_negative'
  | 'indicator_slope_bullish'
  | 'indicator_slope_bearish'
  | 'indicator_signal_change'
  | 'crsi_band_extremes'

export type AlertTriggerMode =
  | 'once'
  | 'once_per_bar'
  | 'once_per_bar_close'

// =============================================================================
// Guest Alert Storage (localStorage)
// =============================================================================

/**
 * Convert AlertCreate to GuestAlert for localStorage storage.
 * Maps between API format and guest format.
 */
function toGuestAlert(alert: AlertCreate, symbol: string): GuestAlert {
  // Map API conditions to guest conditions
  const conditionMap: Record<string, GuestAlertCondition> = {
    'above': 'above',
    'below': 'below',
    'crosses_up': 'crosses-up',
    'crosses_down': 'crosses-down',
    // Indicator conditions map to 'above' for guest mode (simplified)
    'indicator_crosses_upper': 'above',
    'indicator_crosses_lower': 'below',
    'indicator_turns_positive': 'above',
    'indicator_turns_negative': 'below',
    'indicator_slope_bullish': 'above',
    'indicator_slope_bearish': 'below',
    'indicator_signal_change': 'above',
  }

  const now = new Date().toISOString()

  return {
    uuid: crypto.randomUUID(),
    symbol,
    condition: conditionMap[alert.condition] || 'above',
    target: alert.threshold ?? 0,
    enabled: alert.is_active ?? true,
    created_at: now,
    updated_at: now,
  }
}

/**
 * Save guest alert to localStorage.
 * Called when user is in guest mode (not authenticated).
 */
export function saveGuestAlert(alert: AlertCreate, symbol: string): GuestAlert {
  const STORAGE_KEY = 'polishedcharts_data'

  // Get existing localStorage data
  const raw = localStorage.getItem(STORAGE_KEY)
  let data: any = { schemaVersion: 1, alerts: [], watchlist: { uuid: '', symbols: [], sort_order: [], created_at: '', updated_at: '' }, layouts: [] }

  if (raw) {
    try {
      data = JSON.parse(raw)
    } catch (e) {
      console.error('Failed to parse localStorage:', e)
    }
  }

  // Ensure schema structure
  if (!data.alerts) data.alerts = []
  if (!data.schemaVersion) data.schemaVersion = 1

  // Create guest alert
  const guestAlert = toGuestAlert(alert, symbol)

  // Add to alerts array
  data.alerts.push(guestAlert)
  data.updated_at = new Date().toISOString()

  // Save back to localStorage
  localStorage.setItem(STORAGE_KEY, JSON.stringify(data))

  return guestAlert
}

/**
 * Get guest alerts from localStorage.
 */
export function getGuestAlerts(): GuestAlert[] {
  const STORAGE_KEY = 'polishedcharts_data'
  const raw = localStorage.getItem(STORAGE_KEY)

  if (!raw) return []

  try {
    const data = JSON.parse(raw)
    return data.alerts || []
  } catch (e) {
    console.error('Failed to parse guest alerts:', e)
    return []
  }
}

/**
 * Delete guest alert from localStorage by UUID.
 */
export function deleteGuestAlert(uuid: string): void {
  const STORAGE_KEY = 'polishedcharts_data'
  const raw = localStorage.getItem(STORAGE_KEY)

  if (!raw) return

  try {
    const data = JSON.parse(raw)
    if (data.alerts) {
      data.alerts = data.alerts.filter((a: GuestAlert) => a.uuid !== uuid)
      localStorage.setItem(STORAGE_KEY, JSON.stringify(data))
    }
  } catch (e) {
    console.error('Failed to delete guest alert:', e)
  }
}

/**
 * Alert with all backend fields including TradingView-style alert features
 */
export interface Alert {
  id: number
  symbol_id: number
  symbol_ticker?: string  // Added for frontend display
  symbol?: string  // Computed from symbol_ticker for frontend compatibility
  condition: string
  threshold: number | null
  is_active: boolean
  interval?: string  // Timeframe: '1d', '1h', '15m', etc.
  cooldown?: number
  created_at: string
  last_triggered_at?: string | null
  // Trigger mode: once, once_per_bar, once_per_bar_close
  trigger_mode?: AlertTriggerMode
  // Computed label (e.g., "cRSI(20) band cross")
  alert_label?: string
  // Indicator fields
  indicator_name?: string | null
  indicator_field?: string | null
  indicator_params?: Record<string, number | string> | null
  // Direction-specific trigger messages (legacy cRSI format)
  message_upper?: string
  message_lower?: string
  // Flexible messages map (new format)
  messages?: Record<string, string>
  // Enabled conditions (maps condition_type to enabled state)
  enabled_conditions?: Record<string, boolean>
}

/**
 * Alert trigger event with trigger_type and trigger_message
 */
export interface AlertTrigger {
  id: number
  alert_id: number
  triggered_at: string
  observed_price: number | null
  indicator_value?: number | null
  // Direction-specific tracking
  trigger_type?: 'upper' | 'lower'  // Which condition fired
  trigger_message?: string           // Message used for this trigger
  alert_label?: string               // Computed label for display
  delivery_status?: string
  retry_count?: number
  last_retry_at?: string
}

/**
 * Form data for creating a new indicator alert
 */
export interface IndicatorAlertFormData {
  symbol: string
  indicator_name: string
  indicator_field: string
  indicator_params: Record<string, number | string>
  // Flexible enabled conditions - maps condition_type to enabled state
  // e.g., { "indicator_crosses_upper": true, "indicator_crosses_lower": true }
  // or { "indicator_turns_positive": true, "indicator_turns_negative": false }
  enabled_conditions: Record<string, boolean>
  // Flexible messages - maps condition_type to message
  // e.g., { "indicator_crosses_upper": "Time to sell!", "indicator_crosses_lower": "Time to buy!" }
  messages: Record<string, string>
  // Cooldown in seconds (default: 60, minimum: 5)
  cooldown: number
}

export interface AlertCreate {
  symbol?: string           // Symbol ticker (e.g., "AAPL") - preferred, will be resolved by backend
  symbol_id?: number        // Numeric ID - alternative to symbol
  condition: AlertCondition | string
  threshold?: number
  is_active?: boolean
  interval?: string         // Timeframe: '1d', '1h', '15m', '5m', '1m', etc.
  cooldown?: number
  trigger_mode?: AlertTriggerMode  // Trigger mode: once, once_per_bar, once_per_bar_close
  // Indicator fields
  indicator_name?: string
  indicator_field?: string
  indicator_params?: Record<string, number | string>
  // Direction-specific trigger messages (legacy cRSI format)
  message_upper?: string
  message_lower?: string
  // Flexible messages map (new format - takes priority)
  messages?: Record<string, string>
  // Enabled conditions (maps condition_type to enabled state)
  enabled_conditions?: Record<string, boolean>
}

export interface AlertUpdate {
  condition?: AlertCondition | string
  threshold?: number
  is_active?: boolean
  cooldown?: number
  trigger_mode?: AlertTriggerMode
  indicator_name?: string
  indicator_field?: string
  indicator_params?: Record<string, number | string>
  message_upper?: string
  message_lower?: string
  messages?: Record<string, string>
  enabled_conditions?: Record<string, boolean>
}

/**
 * Response from GET /api/v1/alerts/indicator-conditions
 */
export interface IndicatorConditionResponse {
  indicator_name: string
  conditions: IndicatorCondition[]
}

/**
 * Single indicator condition template
 */
export interface IndicatorCondition {
  condition_type: string
  label: string
  description: string
  requires_threshold: boolean
  applicable_fields: string[]
}

export const createAlert = async (alert: AlertCreate): Promise<Alert> => {
  const response = await client.post<Alert>('/alerts/', alert)
  return response.data
}

/**
 * List alerts with optional symbol and active status filtering
 * @param symbol - Symbol string (e.g., "AAPL") - preferred over symbolId
 * @param symbolId - Symbol ID (numeric) - alternative to symbol
 * @param isActive - Filter by active status
 */
export const listAlerts = async (options?: {
  symbol?: string
  symbolId?: number
  isActive?: boolean
}): Promise<Alert[]> => {
  const params: Record<string, any> = {}
  if (options?.symbol !== undefined) params.symbol = options.symbol
  if (options?.symbolId !== undefined) params.symbol_id = options.symbolId
  if (options?.isActive !== undefined) params.is_active = options.isActive
  const response = await client.get<Alert[]>('/alerts/', { params })
  return response.data
}

// Alias for backward compatibility
export const getAlerts = listAlerts

export const getAlert = async (alertId: number): Promise<Alert> => {
  const response = await client.get<Alert>(`/alerts/${alertId}`)
  return response.data
}

export const updateAlert = async (alertId: number, updates: AlertUpdate): Promise<Alert> => {
  const response = await client.put<Alert>(`/alerts/${alertId}`, updates)
  return response.data
}

/**
 * Mute an alert (prevent it from creating trigger events)
 */
export const muteAlert = async (alertId: number): Promise<Alert> => {
  const response = await client.post<Alert>(`/alerts/${alertId}/mute`)
  return response.data
}

/**
 * Unmute an alert (allow it to create trigger events again)
 */
export const unmuteAlert = async (alertId: number): Promise<Alert> => {
  const response = await client.post<Alert>(`/alerts/${alertId}/unmute`)
  return response.data
}

export const deleteAlert = async (alertId: number): Promise<void> => {
  await client.delete(`/alerts/${alertId}`)
}

export const getAlertTriggers = async (alertId: number, limit: number = 100): Promise<AlertTrigger[]> => {
  const response = await client.get<AlertTrigger[]>(`/alerts/${alertId}/triggers`, {
    params: { limit }
  })
  return response.data
}

/**
 * Get recent alert triggers for the global Log tab
 * @param options - Filtering options
 */
export const getRecentTriggers = async (options?: {
  symbol?: string
  symbolId?: number
  limit?: number
  offset?: number
}): Promise<AlertTrigger[]> => {
  const params: Record<string, any> = {
    limit: options?.limit ?? 500,
    offset: options?.offset ?? 0,
  }
  if (options?.symbol !== undefined) params.symbol = options.symbol
  else if (options?.symbolId !== undefined) params.symbol_id = options.symbolId
  const response = await client.get<AlertTrigger[]>('/alerts/triggers/recent', { params })
  return response.data
}

/**
 * Delete a specific alert trigger (log entry)
 * DELETE /api/v1/alerts/triggers/{trigger_id}
 */
export const deleteTrigger = async (triggerId: number): Promise<void> => {
  await client.delete(`/alerts/triggers/${triggerId}`)
}

/**
 * Get available alert conditions for a specific indicator
 * GET /api/v1/alerts/indicator-conditions?indicator_name=crsi
 */
export const getIndicatorConditions = async (indicatorName: string): Promise<IndicatorConditionResponse> => {
  const response = await client.get<IndicatorConditionResponse>('/alerts/indicator-conditions', {
    params: {
      indicator_name: indicatorName,
      _t: Date.now()  // Cache-busting timestamp
    }
  })
  return response.data
}

/**
 * Update notification settings for an alert
 * PATCH /api/v1/notifications/alert-settings/{alert_id}
 */
export const updateAlertNotificationSettings = async (
  alertId: string,
  settings: AlertNotificationSettingsUpdate
): Promise<void> => {
  const authClient = await createAuthenticatedAxios()
  await authClient.patch(`/notifications/alert-settings/${alertId}`, settings)
}

/**
 * Get notification settings for an alert
 * GET /api/v1/notifications/alert-settings/{alert_id}
 */
export const getAlertNotificationSettings = async (alertId: string): Promise<AlertNotificationSettingsUpdate> => {
  const authClient = await createAuthenticatedAxios()
  const response = await authClient.get(`/notifications/alert-settings/${alertId}`)
  return response.data
}
